python/cuml/dask/linear_model/linear_regression.py

```diff
@@ -49,512 +46,157 @@ def __init__(self, fit_intercept=True, normalize=False):
-
-    arrs, dev = arr
-    ret = [(X.get_ipc_handle(),
-            y.get_ipc_handle(),
-            coef.get_ipc_handle()) for X, y, coef in arrs]
-
-    return ret, dev
-
-
-def as_gpu_matrix(arr):
-    blap = arr.compute()
-    mat = blap.as_gpu_matrix(order="F")
-    dev = device_of_devicendarray(mat)
-
-    # Return canonical device id as string
-    return mat, dev
-
-
-def to_gpu_array(arr):
-
-    mat = arr.to_gpu_array()
-    dev = device_of_devicendarray(mat)
-
-    # Return canonical device id as string
-    return mat, dev
-
-
-def fit_to_device_arrays(arr):
-    """
-    :param arr:
-        A tuple in the form of (X, y, coef)
-    :return:
-    """
-
-    mats = [(X.compute().as_gpu_matrix(order='F'),
-             y.to_gpu_array(),
-             coef) for X, y, coef in arr]
-
-    dev = device_of_devicendarray(mats[0][0])
-
-    # Return canonical device id as string
-    return mats, dev
-
-
-def predict_to_device_arrays(arr, worker, loc_dict, nparts, dtype):
-    """
-    :param arr:
-        A tuple in the form of (X, y, coef)
-    :return:
-    """
-    part_number = list(loc_dict.keys())[list(loc_dict.values()).index(worker)]
-
-    mats = []
-    for X, coef in arr:
-        nrows = len(X)
-        part_size = ceil(nrows / nparts)
-        up_limit = min((part_number+1)*part_size, nrows)
-        mat = X.compute().as_gpu_matrix(order='F')
-        pred = rmm.to_device(np.zeros(up_limit-(part_number*part_size),
-                                      dtype=dtype))
-        mats.append([mat, coef.to_gpu_array(), pred])
-
-    dev = device_of_devicendarray(mats[0][0])
-
-    # Return canonical device id as string
-    return mats, dev
-
-
-def extract_part(data, part):
-    return data[part]
-
-
-def preprocess_on_worker(arr):
-    return arr
-
-
-def dev_array_on_worker(rows, dtype=np.float64, unique=0):
-    return rmm.to_device(np.zeros(rows, dtype=dtype))
-
-
-# Need to have different named function for predict to avoid
-# dask key colision in case of same rows and columns between
-# different arrays
-def pred_array_on_worker(rows, cols, dtype=np.float64, unique=0):
-    return rmm.to_device(np.zeros((rows, cols), dtype=dtype))
-
-
-def preprocess_predict(arr):
-    return arr
-
-
-def series_on_worker(ary, worker, loc_dict, nparts, X):
-    nrows = len(X)
-    part_number = list(loc_dict.keys())[list(loc_dict.values()).index(worker)]
-    part_size = ceil(nrows / nparts)
-    up_limit = min((part_number+1)*part_size, nrows)
-    if len(ary) == nparts - 1:
-        idx = (part_number*len(ary[0][0][2]),
-               (part_number+1)*len(ary[0][0][2]))
-    else:
-        idx = (up_limit-len(ary[0][0][2]), up_limit)
-
-    ret = cudf.Series(ary[0][0][2], index=cudf.core.index.RangeIndex(idx[0],
-                                                                     idx[1]))
-    return ret
-
-
-def get_meta(df):
-    ret = df.iloc[:0]
-    return ret
-
+        gpu_futures = self.client.sync(extract_ddf_partitions, X)
```

Comment:
We can do simple task based predict when we are able to create multiple streams in python layer. Otherwise all the predict calls will be serialized on GPUs. That's why I am using multiple streams inside C++ layer to utilize better parallelism. There is actually an issue to create streams in python layer.